sycl/test-e2e/Matrix/joint_matrix_down_convert.cpp

```diff
@@ -0,0 +1,110 @@
+  buffer<bfloat16, 2> bufA(A.get_data(), range<2>(M, K));
+  buffer<float, 2> bufC((float *)C.get_data(), range<2>(M, N));
+
+  queue q;
+  q.submit([&](handler &cgh) {
+     auto accC = bufC.get_access<access::mode::read_write>(cgh);
+     auto accA = bufA.get_access<access::mode::write>(cgh);
+
+     cgh.parallel_for(
+         nd_range<2>({NDRangeM, NDRangeN * SG_SZ}, {1, 1 * SG_SZ}),
+         [=](nd_item<2> spmd_item) [[intel::reqd_sub_group_size(SG_SZ)]] {
+           // The submatrix API has to be accessed by all the workitems in a
+           // subgroup these functions will be called once by the subgroup no
+           // code divergence between the workitems
+           const auto global_idx = spmd_item.get_global_id(0);
+           const auto global_idy = spmd_item.get_global_id(1);
+           const auto sg_startx = global_idx - spmd_item.get_local_id(0);
+           const auto sg_starty = global_idy - spmd_item.get_local_id(1);
+
+           sub_group sg = spmd_item.get_sub_group();
+           joint_matrix<sub_group, bfloat16, use::a, TM, TK, layout::row_major>
+               sub_a;
+           joint_matrix<sub_group, float, use::accumulator, TM, TN> sub_c;
+
+           joint_matrix_load(
+               sg, sub_c,
+               accC.template get_multi_ptr<access::decorated::no>() +
+                   (sg_startx * TM) * N + sg_starty / SG_SZ * TN,
+               N, layout::row_major);
+           // This will be replaced by joint_matrix_copy API
+           // joint_matrix_copy(sg, sub_c, sub_ac);
+           auto wi_slice_c =
+               sycl::ext::intel::experimental::matrix::get_wi_data(sg, sub_c);
+           auto wi_slice_a =
+               sycl::ext::intel::experimental::matrix::get_wi_data(sg, sub_a);
+           for (int i = 0; i < wi_slice_c.length(); i++) {
+             wi_slice_a[i] = (bfloat16)wi_slice_c[i];
+           }
+           ext::intel::experimental::matrix::joint_matrix_store(
+               sg, sub_a,
+               accA.template get_multi_ptr<access::decorated::no>() +
+                   (sg_startx * TM) * N + sg_starty / SG_SZ * TN,
+               N);
+         }); // parallel for
+   }).wait();
+}
+
+int main() {
+  static constexpr size_t MATRIX_M = TM * 2;
+  static constexpr size_t MATRIX_N = TN * 2;
+  static constexpr size_t MATRIX_K = TK * 2;
+  bfloat16 A[MATRIX_M][MATRIX_K];
+  float C[MATRIX_M][MATRIX_N];
+  std::random_device dev;
+
+  std::uniform_real_distribution<float> fdistr(-1.0, 1.0);
+
+  for (int i = 0; i < MATRIX_M; i++) {
+    for (int j = 0; j < MATRIX_N; j++) {
+      C[i][j] = fdistr(dev);
+    }
+  }
+
+  big_matrix<float, MATRIX_M, MATRIX_N> MC((float *)&C);
+  big_matrix<bfloat16, MATRIX_M, MATRIX_K> MA((bfloat16 *)&A);
+  matrix_copy(MC, MA);
+
+  bool res = true;
+  for (int i = 0; i < MATRIX_M; i++) {
+    for (int j = 0; j < MATRIX_N; j++) {
+      if (fabs(A[i][j] - (bfloat16)C[i][j]) > BF16_EPSILON) {
+        res = false;
+      }
+    }
+  }
```

Comment:
Can be replaced with matrix_compare from common.hpp I suggested here: https://github.com/intel/llvm/pull/9845

Reply:
We can add one more template parameter to be typename T1, typename T2 and modify verification code like below:
`if ((fabs(src[i * cols + j] - (T1)ref[i * cols + j])) > BF16_EPSILON) {`
So the function will cover both cases - when conversion needed and when not needed.
I think we need to put it in common, since algorithm is the same and if there are errors, we would need to fix the same code everywhere.